### PR TITLE
release conflicted ip of stateless workload to trigger assigning a new one

### PR DESCRIPTION
**Pull request by Icarus9913**
Once starting a stateless Pod, the spiderpool will reuse the first allocated IPs from the SpiderEndpoint resource for the Pod if it met some errors. If the coordinator detected the IP conflict, spiderpool will still reuse the conflicted IPs. With this, the Pod would never starts.

So, I added a feature that let coordinator send a signal to spiderpool-agent component to release its whole NICs IPs and let the spiderpool reallocate IPs for the Pod. It might let the Pod starts correctly.

Signed-off-by: Icarus9913 [icaruswu66@qq.com](mailto:icaruswu66@qq.com)

### CODE REVIEW

**Comment by weizhoublue** [comment]
这应该在 相关 doc 中有说明：对于 无状态 pod，XXXX

File: docs/concepts/coordinator-zh_CN.md [hunk 1]
```diff
@@ -57,7 +57,7 @@ ClusterIP 的路由，导致无法访问。
 ## 支持检测 Pod 的 IP 是否冲突( alpha 阶段)
 
 对于 Underlay 网络，IP 冲突是无法接受的，这可能会造成严重的问题。在创建 Pod 时，我们可借助 `coordinator` 检测 Pod 的 IP 是否冲突，支持同时检测 IPv4 和 IPv6 地址。通过发送 ARP 或 NDP 探测报文，
-如果发现回复报文的 Mac 地址不是 Pod 本身，那我们认为这个 IP 是冲突的，并拒绝 IP 冲突的 Pod 被创建:
+如果发现回复报文的 Mac 地址不是 Pod 本身，那我们认为这个 IP 是冲突的，并拒绝 IP 冲突的 Pod 被创建。此外，我们默认还会为**无状态**的 Pod 释放所有的已分配的 IP 使其重新分配，您可通过 spiderpool-agent [环境变量](../reference/spiderpool-agent.md#env) `SPIDERPOOL_ENABLED_RELEASE_CONFLICT_IPS` 来控制此功能。
```
Comment (weizhoublue) [inline docs/concepts/coordinator-zh_CN.md:60]: 这句话新加的 可以 描述的更加精准，为什么设计不支持 st，这些都是设计理念要呈现给 用户 。
对于发生IP冲突的**无状态**的 Pod，Spiderpool 支持释放其本轮分配的 冲突 IP 地址，使得 POD 在下一次重建时能够尝试分配到其它非冲突的 IP 。而对于
对于发生IP冲突的**有状态**的 Pod，为了保持 IP 地址也是有状态设计，所以并不具备该释放 IP 的设计

File: cmd/coordinator/cmd/command_add.go [hunk 2]
```diff
@@ -197,6 +200,22 @@ func CmdAdd(args *skel.CmdArgs) (err error) {
 
 	if err = errg.Wait(); err != nil {
 		logger.Error("failed to detect gateway and ip checking", zap.Error(err))
+		if errors.Is(err, constant.ErrIPConflict) {
+			_, innerErr := client.Daemonset.DeleteIpamIps(daemonset.NewDeleteIpamIpsParams().WithContext(context.TODO()).WithIpamBatchDelArgs(
```
Comment (weizhoublue) [inline cmd/coordinator/cmd/command_add.go:204]: is there log when succeeding to release ?

File: cmd/spiderpool-agent/cmd/config.go [hunk 3]
```diff
@@ -59,6 +59,7 @@ var envInfo = []envConf{
 	{"SPIDERPOOL_METRIC_HTTP_PORT", "5711", true, &agentContext.Cfg.MetricHttpPort, nil, nil},
 	{"SPIDERPOOL_GOPS_LISTEN_PORT", "5712", false, &agentContext.Cfg.GopsListenPort, nil, nil},
 	{"SPIDERPOOL_PYROSCOPE_PUSH_SERVER_ADDRESS", "", false, &agentContext.Cfg.PyroscopeAddress, nil, nil},
+	{"SPIDERPOOL_ENABLED_RELEASE_CONFLICT_IPS", "true", true, nil, &agentContext.Cfg.EnableReleaseConflictIPs, nil},
```
Comment (weizhoublue) [inline cmd/spiderpool-agent/cmd/config.go:62]: EnableReleaseConflictIPsForStateless

File: docs/concepts/coordinator.md [hunk 4]
```diff
@@ -57,7 +57,8 @@ For more information about the Underlay Pod not being able to access the Cluster
 ## Detect Pod IP conflicts(alpha)
 
 IP conflicts are unacceptable for underlay networks, which can cause serious problems. When creating a pod, we can use the `coordinator` to detect whether the IP of the pod conflicts, and support both IPv4 and IPv6 addresses. By sending an ARP or NDP probe message,
-If the MAC address of the reply packet is not the pod itself, we consider the IP to be conflicting and reject the creation of the pod with conflicting IP addresses:
+If the MAC address of the reply packet is not the pod itself, we consider the IP to be conflicting and reject the creation of the pod with conflicting IP addresses.
```
Comment (weizhoublue) [inline docs/concepts/coordinator.md:60]: 这个翻译过来的有点 中式
If the MAC address of the reply packet is belonged to the pod itself, the IP address is recognized as conflicted .... 
@windsonsea 给一些建议

File: pkg/ipam/release.go [hunk 5]
```diff
@@ -204,3 +206,72 @@ func (i *ipam) release(ctx context.Context, uid string, details []spiderpoolv2be
...
 	return nil
 }
+
+// ReleaseIPs will release the given IP corresponding NIC whole IPs,
+// and we will release the SpiderEndpoint recorded IPs first and release the SpiderIPPool recorded IPs later.
+func (i *ipam) ReleaseIPs(ctx context.Context, delArgs *models.IpamBatchDelArgs) (err error) {
+	log := logutils.FromContext(ctx)
+
+	var pod *corev1.Pod
+	// with IP release, we must need Pod UID
```
Comment (weizhoublue) [inline pkg/ipam/release.go:216]: it needs to offer Pod UID when releasing ip

File: pkg/ipam/release.go [hunk 6]
```diff
@@ -204,3 +206,72 @@ func (i *ipam) release(ctx context.Context, uid string, details []spiderpoolv2be
...
+					return fmt.Errorf("failed to get Pod '%s/%s', error: %v", *delArgs.PodNamespace, *delArgs.PodName, err)
+				}
+			}
+
+			podTopController, err := i.podManager.GetPodTopController(ctx, pod)
+			if nil != err {
+				return fmt.Errorf("failed to get the top controller of the Pod %s/%s, error: %v", pod.Namespace, pod.Name, err)
+			}
+
+			// do not release conflict IPs for stable Pod
```
Comment (weizhoublue) [inline pkg/ipam/release.go:245]: do not release conflict IPs for stateful Pod

File: pkg/ipam/release.go [hunk 7]
```diff
@@ -204,3 +206,72 @@ func (i *ipam) release(ctx context.Context, uid string, details []spiderpoolv2be
...
+			if (i.config.EnableStatefulSet && podTopController.APIVersion == appsv1.SchemeGroupVersion.String() && podTopController.Kind == constant.KindStatefulSet) ||
```
Comment (weizhoublue) [inline pkg/ipam/release.go:246]: how about kubevirt, I remember kubevirt is supposed to be stateful ?

File: pkg/ipam/release.go [hunk 8]
```diff
@@ -204,3 +206,72 @@ func (i *ipam) release(ctx context.Context, uid string, details []spiderpoolv2be
...
+				(i.config.EnableKubevirtStaticIP && podTopController.APIVersion == kubevirtv1.SchemeGroupVersion.String() && podTopController.Kind == constant.KindKubevirtVMI) {
+				log.Warn("no need to release conflict IPs for stable Pod")
```
Comment (weizhoublue) [inline pkg/ipam/release.go:248]: no need to release conflict IPs for stateful Pod , pod uid ***** , ip ****

File: pkg/ipam/release.go [hunk 9]
```diff
@@ -204,3 +206,72 @@ func (i *ipam) release(ctx context.Context, uid string, details []spiderpoolv2be
...
+				return nil
+			}
+		} else {
+			log.Warn("EnableReleaseConflictIPs is disabled, skip to release IPs")
```
Comment (weizhoublue) [inline pkg/ipam/release.go:252]: is there IP information in the log ?

File: pkg/ipam/release.go [hunk 10]
```diff
@@ -204,3 +206,72 @@ func (i *ipam) release(ctx context.Context, uid string, details []spiderpoolv2be
...
+			return nil
+		}
+	}
+
+	// release SpiderEndpoint IPs
```
Comment (weizhoublue) [inline pkg/ipam/release.go:257]: release IP for stateless endpoint

File: pkg/ipam/release.go [hunk 11]
```diff
@@ -204,3 +206,72 @@ func (i *ipam) release(ctx context.Context, uid string, details []spiderpoolv2be
...
+	}
+	recordedIPAllocationDetails, err := i.endpointManager.ReleaseEndpointIPs(ctx, endpoint, *delArgs.PodUID)
+	if nil != err {
+		return fmt.Errorf("failed to release SpiderEndpoint IPs, error: %v", err)
+	}
+
+	// release IPPool IPs
+	if len(recordedIPAllocationDetails) != 0 {
+		log.Sugar().Infof("try to release IPs: %v", recordedIPAllocationDetails)
+		err := i.release(ctx, *delArgs.PodUID, recordedIPAllocationDetails)
```
Comment (weizhoublue) [inline pkg/ipam/release.go:270]: has it been tested when multiple interfaces case ?

File: api/v1/agent/openapi.yaml [hunk 12]
```diff
@@ -88,6 +88,12 @@ paths:
         Delete multiple ip for a pod, case for spiderflat compent
       tags:
         - daemonset
+      parameters:
+        - name: ipam-batch-del-args
+          in: body
+          required: true
+          schema:
+            $ref: "#/definitions/IpamBatchDelArgs"
```
Comment (weizhoublue) [inline api/v1/agent/openapi.yaml:96]: for the response , I suggest to be the following, so the binary log is available for telling the detail as well :
500: internal failure
501: forbide to release IP for stateless endpoint
502: forbide to release IP for stateful endpoint

File: cmd/coordinator/cmd/command_add.go (hunk 2 above)
Comment (Icarus9913) [reply cmd/coordinator/cmd/command_add.go:204]: 202行已有日志展示IP冲突的地址以及网卡名。

204行发送请求后，spiderpool-agent有日志输出，以及判断是否为无状态应用

File: pkg/ipam/release.go (hunk 8 above)
Comment (Icarus9913) [reply pkg/ipam/release.go:248]: log初始化的时配置了attributes，pod相关信息已有

File: pkg/ipam/release.go (hunk 10 above)
Comment (Icarus9913) [reply pkg/ipam/release.go:257]: 该接口为统一接口，不区分是否为stateless

File: pkg/ipam/release.go (hunk 11 above)
Comment (Icarus9913) [reply pkg/ipam/release.go:270]: 已测试多网卡，e2e已覆盖

File: api/v1/agent/openapi.yaml (hunk 12 above)
Comment (Icarus9913) [reply api/v1/agent/openapi.yaml:96]: `ReleaseEndpointIPs(ctx context.Context, endpoint *spiderpoolv2beta1.SpiderEndpoint, uid string) ([]spiderpoolv2beta1.IPAllocationDetail, error)`

SpiderEndpoint中ReleaseIPs接口为统一接口，判断是否为 "stateless/stateful" 为调用层判断。若静态Pod则不释放IP并配合Warning级别日志，return 200并不会报错。

Coordinator处，不论释放IP成功与否，都会以“检测到IP冲突”报错return

File: test/e2e/coordinator/macvlan-overlay-one/macvlan_overlay_one_test.go [hunk 13]
```diff
@@ -668,6 +691,172 @@ var _ = Describe("MacvlanOverlayOne", Label("overlay", "one-nic", "coordinator")
...
+			}
+			if frame.Info.IpV6Enabled {
+				err := frame.KClient.Delete(ctx, &conflictV6Pool)
+				Expect(err).NotTo(HaveOccurred())
+
+				commandV6Str := fmt.Sprintf("ip addr del %s dev eth0", firstConflictV6IP)
+				output, err := frame.DockerExecCommand(ctx, common.VlanGatewayContainer, commandV6Str)
+				Expect(err).NotTo(HaveOccurred(), "Failed to exec %s for Node %s, error is: %v, log: %v", commandV6Str, common.VlanGatewayContainer, err, string(output))
+			}
+		})
```
Comment (weizhoublue) [inline test/e2e/coordinator/macvlan-overlay-one/macvlan_overlay_one_test.go:859]: need a case for stateful pod whose IP should not be released ?

File: pkg/ipam/release.go (hunk 10 above)
Comment (weizhoublue) [reply pkg/ipam/release.go:257]: I seems to be dedicated for stateless endpoint after line 257 , so I mean that the comment could be the following, and help code reviewer a lot 

// release IP for stateless endpoint

File: api/v1/agent/openapi.yaml (hunk 12 above)
Comment (weizhoublue) [reply api/v1/agent/openapi.yaml:96]: so I expect it to be better, from the agent return code, the binary log could tell the details

**Review by weizhoublue — DISMISSED** [review]

**Review by ty-dc — APPROVED** [review]

**Pull request merged**